Snowplow/SPEmitter.m

```diff
@@ -172,27 +202,81 @@ - (void) sendEvents {
+                    
+                    // Reset the STM as it is going individually
+                    [eventPayload setValue:[NSString stringWithFormat:@"%ld", (long)[SPUtilities getTimestamp]] forKey:kSPSentTimestamp];
+                    
+                    // Build and Send the event!
+                    [singleEventArray addObject:eventPayload];
+                    [singleIndexArray addObject:[listValues[j] objectForKey:@"ID"]];
+                    SPSelfDescribingJson *payload = [[SPSelfDescribingJson alloc] initWithSchema:kSPPayloadDataSchema
+                                                                                         andData:singleEventArray];
+                    [self sendEventWithRequest:[self getRequestPostWithData:payload] andIndex:singleIndexArray andResultArray:sendResults andOversize:YES];
+                } else if (totalByteSize + payloadByteSize > _byteLimitPost) {
+                    // Adding this event to the accumulated array would exceed the limit.
+                    SPSelfDescribingJson *payload = [[SPSelfDescribingJson alloc] initWithSchema:kSPPayloadDataSchema
+                                                                                         andData:eventArray];
+                    [self sendEventWithRequest:[self getRequestPostWithData:payload] andIndex:indexArray andResultArray:sendResults andOversize:NO];
+                    
+                    // Reset collections and STM
+                    eventArray = [[NSMutableArray alloc] init];
+                    indexArray = [[NSMutableArray alloc] init];
+                    stm = [SPUtilities getTimestamp];
+                    
+                    // Add new STM to the payload
+                    [eventPayload setValue:[NSString stringWithFormat:@"%ld", (long)stm] forKey:kSPSentTimestamp];
+                    
+                    // Add event to collections
+                    [eventArray addObject:eventPayload];
+                    [indexArray addObject:[listValues[j] objectForKey:@"ID"]];
+                    
+                    // Update byte count
+                    totalByteSize = payloadByteSize;
+                } else {
+                    // Add event to collections
+                    [eventArray addObject:eventPayload];
+                    [indexArray addObject:[listValues[j] objectForKey:@"ID"]];
+                    
+                    // Update byte count
+                    totalByteSize += payloadByteSize;
+                }
             }
             
-            NSMutableDictionary *payload = [[NSMutableDictionary alloc] init];
-            [payload setValue:kSPPayloadDataSchema forKey:@"schema"];
-            [payload setValue:eventArray forKey:@"data"];
-            [self sendSyncRequest:[self getRequestPostWithData:payload] withIndex:indexArray withResultPointer:sendResults];
+            // If we have not sent all of the events...
+            if (eventArray.count > 0) {
```

Comment:
Should we create the stm here instead of at the start of the for loop? If we have 1 small event and 1 huge event then the sequence of events with the current will be (I think):

* new stm attached to small event
* big event found to be over the byte limit
* new stm attached to big event
* big event sent alone (possible large delay here)
* small event sent alone

Then the small event could end up being sent significantly later than its stm was generated.